### PR TITLE
updated customer-order

### PR DESCRIPTION
**Pull request by Asadupramod012**
updated customer-order with required JSON format.

### CODE REVIEW

File: apps/customer/urls.py [hunk 1]
```diff
@@ -12,5 +12,6 @@
 
 urlpatterns = [
     path('', include(router.urls)),
-    
+    path('customer-order/', TestCustomerCreateViews.as_view(), name='customer-order-create'),
+    path('customer-order/<str:pk>/', TestCustomerCreateViews.as_view(), name='customer-order-details'),
```
Comment (QodeFarm) [inline apps/customer/urls.py:16]: Why it is customer-order it should be customers/

File: apps/customer/views.py [hunk 2]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
     def update(self, request, *args, **kwargs):
-        return update_instance(self, request, *args, **kwargs)
\ No newline at end of file
+        return update_instance(self, request, *args, **kwargs)
+    
+#==========================================================================
+
+
+def add_key_value_to_all_ordereddicts(od_list, key, value):
+    for od in od_list:
```
Comment (QodeFarm) [inline apps/customer/views.py:90]: what is od and od_list here, the variable name should be more understandable

File: apps/customer/views.py [hunk 3]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+        filter_kwargs = {main_model_pk_field_name: del_value}
+ 
+    deleted_count, _ = current_model_name.objects.filter(**filter_kwargs).delete()
+ 
+    if deleted_count > 0:
+        print(f'***Data Deleted Successfully***')
+    else:
+        return Response({f'***error: {current_model_name} not found or already deleted.***'})    
+
+class TestCustomerCreateViews(generics.CreateAPIView,mi.ListModelMixin,mi.CreateModelMixin,mi.RetrieveModelMixin,mi.UpdateModelMixin,mi.DestroyModelMixin):
```
Comment (QodeFarm) [inline apps/customer/views.py:136]: Why the class name is starting with Test? It should not.

File: apps/customer/views.py [hunk 4]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+        pk = request.data['customers'].get('customer_id')
+        update_multi_instance(customer_attachments_data,pk,Customer,CustomerAttachments,CustomerAttachmentsSerializers)
+        print('***CustomerAttachements updated ***')
+
+        customer_addresses_data = request.data.pop('customer_addresses')
+        pk = request.data['customers'].get('customer_id')
+        update_multi_instance(customer_addresses_data,pk,Customer,CustomerAddresses,CustomerAddressesSerializers,main_model_field_name='customer_id')
+        print('***CustomerAddresses updated ***')
+
+        return Response({'***data updated successfully***'})
```
Comment (QodeFarm) [inline apps/customer/views.py:222]: response should be the updated data and it should match out JSON pattern

File: apps/customer/views.py [hunk 5]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+        attachments_related_data = CustomerAddresses.objects.filter(customer_id=str(customer_id))
+        attachments_related_serializer = CustomerAddressesSerializers(attachments_related_data, many=True)
+
+        # Customizing the response data
+        custom_data = {
+            "customers": serializer.data,
+            "customer_attachments": items_related_serializer.data,
+            "customer_addresses":attachments_related_serializer.data,
+        }
+        return Response(custom_data)
```
Comment (QodeFarm) [inline apps/customer/views.py:199]: response should match our JSON pattern

File: apps/customer/views.py [hunk 6]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+            print(customer_id)
+            add_key_value_to_all_ordereddicts(customer_attachments_data,'customer_id',customer_id)
+            create_multi_instance(customer_attachments_data,CustomerAttachmentsSerializers)
+            print('***Customer-Attachments-successful***')
+
+            add_key_value_to_all_ordereddicts(customer_addresses_data,'customer_id',customer_id)
+            create_multi_instance(customer_addresses_data,CustomerAddressesSerializers)
+            print('***Customer-Addresses-successful***')
+
+            return Response({customer_id})
```
Comment (QodeFarm) [inline apps/customer/views.py:175]: response should be the created data and it should match our JSON pattern

File: apps/customer/views.py [hunk 7]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+        pk = kwargs.get('pk')
+
+        try:
+            instance = Customer.objects.get(pk=pk)
+            del_value = instance.customer_id  # Fetch value from main model
+            delete_multi_instance(del_value,Customer,CustomerAttachments,CustomerAttachmentsSerializers,main_model_field_name='customer_id')
+            delete_multi_instance(del_value,Customer,CustomerAddresses,CustomerAddressesSerializers,main_model_field_name='customer_id')
+            instance.delete()
+
+            return Response({'***Data Deleted Successfully***'}, status=status.HTTP_204_NO_CONTENT)
```
Comment (QodeFarm) [inline apps/customer/views.py:234]: response should match our JSON pattern

File: apps/customer/views.py [hunk 8]
```diff
@@ -81,4 +81,157 @@ def create(self, request, *args, **kwargs):
...
+
+        except Customer.DoesNotExist:
+            return Response({'error': 'Instance not found.'}, status=status.HTTP_404_NOT_FOUND)
```
Comment (QodeFarm) [inline apps/customer/views.py:237]: response should match our JSON pattern

**Review by QodeFarm — COMMENTED** [review]
Remove all the print statements
for GET, POST, PUT, DEL, the response should match our JSON pattern
POST - Should return the newly added data in in our JSON pattern
PUT - Should return the updated data in our JSON pattern
DEL - Should return the 204 No Content in our JSON pattern
In case of any exceptions/failures also the response should match our JSON pattern

File: apps/customer/views.py (hunk 3 above)
Comment (Asadupramod012) [reply apps/customer/views.py:136]: Updated

File: apps/customer/urls.py (hunk 1 above)
Comment (Asadupramod012) [reply apps/customer/urls.py:16]: updated

File: apps/customer/views.py (hunk 5 above)
Comment (Asadupramod012) [reply apps/customer/views.py:199]: updated

File: apps/customer/views.py (hunk 2 above)
Comment (Asadupramod012) [reply apps/customer/views.py:90]: Bhagyaraju Updated this method in utils, so I need to import this after his code merged into main.

File: apps/customer/views.py [hunk 9]
```diff
@@ -81,4 +82,124 @@ def create(self, request, *args, **kwargs):
...
+#==========================================================================
+
+class CustomerCreateViews(generics.CreateAPIView,mi.ListModelMixin,mi.CreateModelMixin,mi.RetrieveModelMixin,mi.UpdateModelMixin,mi.DestroyModelMixin):
+    queryset = Customer.objects.all()
+    serializer_class = CustomerSerializer
+    
+    def get(self, request, *args, **kwargs):
+        if 'pk' in kwargs:
+            return self.retrieve(request, *args, **kwargs)
+        return list_all_objects(self, request, *args, **kwargs)
```
Comment (QodeFarm) [inline apps/customer/views.py:96]: Do not use multiple return statements here, use single return statement.

File: apps/customer/views.py [hunk 10]
```diff
@@ -81,4 +82,124 @@ def create(self, request, *args, **kwargs):
...
+            custom_data = {
+                "customer_data": customer_data,
+                "customer_attachments": attachments_data,
+                "customer_addresses":addresses_data
+            }
+            return Response({
+                'count': 1,
+                'message': 'Record updated successfully',
+                'data': custom_data
+            },status=status.HTTP_200_OK)
```
Comment (QodeFarm) [inline apps/customer/views.py:186]: Use build_response function here

File: apps/customer/views.py [hunk 11]
```diff
@@ -81,4 +82,124 @@ def create(self, request, *args, **kwargs):
...
+        else:
+            return Response({
+                'count': 0,
+                'message': 'Record updation failed',
+                'errors': serializer.errors
+            }, status=status.HTTP_400_BAD_REQUEST)
```
Comment (QodeFarm) [inline apps/customer/views.py:192]: Use build_response function here

Comment (Asadupramod012) [reply apps/customer/views.py]: Updated

File: apps/customer/views.py (hunk 10 above)
Comment (Asadupramod012) [reply apps/customer/views.py]: Updated

File: apps/customer/views.py (hunk 9 above)
Comment (Asadupramod012) [reply apps/customer/views.py]: Updated

**Pull request merged**